### PR TITLE
FFTShift and IFFTShift support

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

File: mo_extensions/front/onnx/fft_ext.py [hunk 1]
```diff
@@ -1,29 +1,81 @@
...
+
+        if not centered:
+            node.in_port(0).get_connection().set_destination(fft.in_port(0))
+            return [fft.id]
+        else:
+            dim = onnx_attr(node, 'dim', 'ints', default=[2, 3])
+
+            shift_i = Const(graph, {'value': [shift_x, shift_y]}).create_node()
+            axes_i = Const(graph, {'value': dim}).create_node()
+            roll_ifftshift = Roll(graph, {'name': node.name + '/ifftshift_'}).create_node()
```
Comment (dkurt) [inline mo_extensions/front/onnx/fft_ext.py:31]: FFT shift should work in this way:
```
    // tl | tr        br | bl
    // ---+---   ->   ---+---
    // bl | br        tr | tl
```

Can you add a test to verify it?

**Comment by dkurt** [comment]
@likholat., can you please update this branch with the changes in C++ implementation with FFTShift?

File: user_ie_extensions/fft_impl.cpp [hunk 2]
```diff
@@ -128,6 +191,9 @@ InferenceEngine::StatusCode FFTImpl::execute(std::vector<InferenceEngine::Blob::
             CvMat* out = cvCreateMatHeader(rows, cols, CV_32FC2);
             cvSetData(inp, reinterpret_cast<void*>(inpData + d * planeSize), cols * 2 * sizeof(float));
             cvSetData(out, reinterpret_cast<void*>(outData + d * planeSize), cols * 2 * sizeof(float));
+
+            fftshift(inp, rows, cols);
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:195]: Please also change test and add shift to it

File: user_ie_extensions/fft_op.cpp [hunk 3]
```diff
@@ -9,13 +9,14 @@ constexpr ngraph::NodeTypeInfo FFTOp::type_info;
...
-FFTOp::FFTOp(const ngraph::Output<ngraph::Node>& inp, bool _inverse) : Op({inp}) {
+FFTOp::FFTOp(const ngraph::Output<ngraph::Node>& inp, bool _inverse, bool _centered) : Op({inp}) {
     constructor_validate_and_infer_types();
     inverse = _inverse;
+    centered = _centered;
 }
 //! [op:ctor]
 
-IFFTOp::IFFTOp(const ngraph::Output<ngraph::Node>& inp, bool _inverse) : FFTOp({inp, true}) {}
+IFFTOp::IFFTOp(const ngraph::Output<ngraph::Node>& inp, bool _inverse, bool _centered) : FFTOp({inp, true, false}) {}
```
Comment (dkurt) [inline user_ie_extensions/fft_op.cpp:19]:
```
IFFTOp::IFFTOp(const ngraph::Output<ngraph::Node>& inp, bool _inverse, bool _centered) : FFTOp({inp, true, _centered}) {}
```

File: user_ie_extensions/cpu_kernel.hpp [hunk 4]
```diff
@@ -40,7 +40,7 @@ class FFTImpl : public InferenceEngine::ILayerExecImpl {
 private:
     ngraph::Shape inpShape;
     ngraph::Shape outShape;
-    bool inverse;
+    bool inverse, centered;
```
Comment (dkurt) [inline user_ie_extensions/cpu_kernel.hpp:43]: Please test this functionality

File: user_ie_extensions/cpu_kernel.hpp [hunk 5]
```diff
@@ -40,7 +40,7 @@ class FFTImpl : public InferenceEngine::ILayerExecImpl {
 private:
```
Comment (dkurt) [inline user_ie_extensions/cpu_kernel.hpp:40]: Change: `.github/workflows/main.yml` with a new wheel version:
```yml
VERSION: 2021.4.2.1
```

File: user_ie_extensions/fft_impl.cpp [hunk 6]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    cvGetRawData(src, (uchar**)&src_data, &src_step, &src_size);
+
+    int h = src_size.height / 2;
+    int w = src_size.width / 2;
+
+    src_step /= sizeof(float);
+    CvMat* tl = new CvMat();
+    CvMat* tr = new CvMat();
+    CvMat* bl = new CvMat();
+    CvMat* br = new CvMat();
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:133]: Do not forget to delete `tl, tr, bl, br` objects. And, probably, `tmp` too.

File: user_ie_extensions/fft_impl.cpp [hunk 7]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+
+    cvInitMatHeader(tl, h, w, CV_32FC2, src_data, src_step * sizeof(float));
+    cvInitMatHeader(tr, h, w, CV_32FC2, src_data + (src_step / 2), src_step * sizeof(float));
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:136]: `src_data + w`?

File: user_ie_extensions/fft_impl.cpp [hunk 8]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    cvInitMatHeader(bl, h, w, CV_32FC2, src_data + h * src_step, src_step * sizeof(float));
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:137]: `src_data + h * src_size.width`

File: user_ie_extensions/fft_impl.cpp [hunk 9]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    cvInitMatHeader(br, h, w, CV_32FC2, src_data + (h * src_step + src_step / 2), src_step * sizeof(float));
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:138]: `src_data + h * src_size.width + w`

File: user_ie_extensions/fft_impl.cpp [hunk 10]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    int h = src_size.height / 2;
+    int w = src_size.width / 2;
+
+    src_step /= sizeof(float);
+    CvMat* tl = new CvMat();
+    CvMat* tr = new CvMat();
+    CvMat* bl = new CvMat();
+    CvMat* br = new CvMat();
+
+    cvInitMatHeader(tl, h, w, CV_32FC2, src_data, src_step * sizeof(float));
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:135]: Do not divide `src_step` by `sizeof(float)` but just use it unchanged:
```cpp
    cvInitMatHeader(tl, h, w, CV_32FC2, src_data, src_step);
```

File: user_ie_extensions/fft_impl.cpp [hunk 11]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    static auto cvCloneMat = reinterpret_cast<cvCloneMatF*>(so->get_symbol("cvCloneMat"));
+    static auto cvCopy = reinterpret_cast<cvCopyF*>(so->get_symbol("cvCopy"));
+    static auto cvInitMatHeader = reinterpret_cast<cvInitMatHeaderF*>(so->get_symbol("cvInitMatHeader"));
+    static auto cvGetRawData = reinterpret_cast<cvGetRawDataF*>(so->get_symbol("cvGetRawData"));
+
+    // tl | tr        br | bl
+    // ---+---   ->   ---+---
+    // bl | br        tr | tl
+
+    float* src_data;
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:121]: just `data`

File: user_ie_extensions/fft_impl.cpp [hunk 12]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    int src_step;
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:122]: `int step`

File: user_ie_extensions/fft_impl.cpp [hunk 13]
```diff
@@ -103,6 +108,45 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+    CvSize src_size;
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:123]: `CvSize size`

File: user_ie_extensions/fft_impl.cpp [hunk 14]
```diff
@@ -128,12 +177,19 @@ InferenceEngine::StatusCode FFTImpl::execute(std::vector<InferenceEngine::Blob::
             CvMat* out = cvCreateMatHeader(rows, cols, CV_32FC2);
             cvSetData(inp, reinterpret_cast<void*>(inpData + d * planeSize), cols * 2 * sizeof(float));
             cvSetData(out, reinterpret_cast<void*>(outData + d * planeSize), cols * 2 * sizeof(float));
+
+            if (centered)
+                fftshift(inp);
+
             if (inverse)
                 cvDFT(inp, out, CV_DXT_INVERSE, 0);
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:185]: Please add a flag with normalization:
```patch
if (inverse)
-  cvDFT(inp, out, CV_DXT_INVERSE, 0);
+  cvDFT(inp, out, CV_DXT_INVERSE_SCALE, 0);
else
-  cvDFT(inp, out, CV_DXT_FORWARD, 0);
+  cvDFT(inp, out, CV_DXT_FORWARD + CV_DXT_SCALE, 0);
- cvScale(out, out, 1.0 / sqrtf(cols * rows), 0);
```

Comment (dkurt) [reply user_ie_extensions/fft_impl.cpp:185]: After all other changes will be checked with DIRECT model

File: user_ie_extensions/fft_impl.cpp (hunk 6 above)
Comment (dkurt) [reply user_ie_extensions/fft_impl.cpp:133]: check `cvReleaseMat`

File: user_ie_extensions/fft_impl.cpp (hunk 14 above)
Comment (likholat) [reply user_ie_extensions/fft_impl.cpp:190]: These changes give the wrong result

File: user_ie_extensions/fft_impl.cpp (hunk 6 above)
Comment (likholat) [reply user_ie_extensions/fft_impl.cpp:136]: ```Segmentation fault (core dumped)```

Comment (dkurt) [reply user_ie_extensions/fft_impl.cpp:136]: What about just `delete tl;` and others?

File: user_ie_extensions/fft_impl.cpp [hunk 15]
```diff
@@ -103,6 +108,50 @@ InferenceEngine::StatusCode FFTImpl::init(InferenceEngine::LayerConfig &config,
...
+
+    float* data;
+    int step;
+    CvSize size;
+    cvGetRawData(src, (uchar**)&data, &step, &size);
+
+    int height = size.height;
+    int width = size.width;
+    int h = height / 2;
+    int w = width / 2;
```
Comment (dkurt) [inline user_ie_extensions/fft_impl.cpp:131]: maybe let's name as `h2 and w2` to distinguish from `height` and `width`

**Pull request merged**